nab/labeler.py

```diff
@@ -206,111 +209,104 @@ def getUserLabels(self):
+      current = None
+      for t in times:
+        if current is None:
+          current = t
+          bucket = [current]
+          continue
+        if (t - current) <= buffer:
+          bucket.append(t)
+        else:
+          rawAnomalies.append(bucket)
+          current = t
+          bucket = [current]
+      if bucket:
+        rawAnomalies.append(bucket)
+
+      # Merge the bucketed timestamps that qualify as true anomalies
+      for bucket in rawAnomalies:
+        if len(bucket) > len(self.userLabels)*self.threshold:
+          trueAnomalies.append(max(bucket, key=bucket.count))
+
+      labels = numpy.array(timestamps.isin(trueAnomalies), dtype=int)
+      combinedLabels[relativePath] = pandas.DataFrame(
+        {"timestamp":timestamps, "label":labels})
+      labelIndices[relativePath] = [i for i in range(len(labels))
+                                    if labels[i]==1]
+  
     self.combinedLabels = combinedLabels
+    self.labelIndices = labelIndices
+    
 
-
-  def combineWindows(self):
-    """Take raw combined labels and compress them to combinedWindows."""
-    combinedWindows = {}
-
-    for relativePath, labels in self.combinedLabels.iteritems():
-      delta = labels["timestamp"][1] - labels["timestamp"][0]
-
-      labels = labels[labels["label"] == 1]
-      dataSetWindows = []
-
-      if labels.shape[0] != 0:
-        t1 = None
-        t0 = None
-
-        for _, row in labels.iterrows():
-          t1 = row["timestamp"]
-
-          if t0 is None:
-            window = [strf(t1)]
-
-          elif t1 - t0 != delta:
-            window.append(strf(t0))
-            dataSetWindows.append(window)
-            window = [strf(t1)]
-
-          t0 = t1
-
-        window.append(strf(t1))
-        dataSetWindows.append(window)
-
-      combinedWindows[relativePath] = dataSetWindows
-
-    self.combinedWindows = combinedWindows
-
-
-  def relaxWindows(self, percentOfDataSet = 0.1):
+  def relaxWindows(self):
     """
-    This takes all windows and relaxes them (expands them) by a certain
-    percentage of the data. A length (relaxWindowLength) is picked beforehand
-    and each window is lengthened on both its left and right side by that
-    length. This length is chosen as a certain percentage of the datafile.
+    This takes all the true anomalies, as calculated by combineRawLabels(), and
+    adds a relaxed window. The window length is the class variable windowSize,
+    and the location is centered on the anomaly timestamp.
     """
     allRelaxedWindows = {}
-
-    for relativePath, limits in self.combinedWindows.iteritems():
-
+    for relativePath, anomalies in self.labelIndices.iteritems():
+    
       data = self.corpus.dataFiles[relativePath].data
-      length = len(data["timestamp"])
-
-      relaxWindowLength = int(percentOfDataSet*length)
+      length = len(data)
+      num = len(anomalies)
+      if num:
+        relaxWindowLength = int(self.windowSize * length / len(anomalies))
```

Comment:
Yes windowSize is a float, but is this a problem? The calculation rounds down to the nearest int, which I think is okay and desired here. This would be a problem for rounding down to 0, but not an issue for us b/c only would happen with really small dataset lengths.